File: src/main/java/nl/brp/api/gebeurtenissen/web/api/GebeurtenisApiDelegateService.java
package nl.brp.api.gebeurtenissen.web.api;

import nl.brp.api.gebeurtenissen.web.validation.Specification;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.util.UUID;
import java.util.List;

@Service
public class GebeurtenisApiDelegateService implements GebeurtenisApiDelegate {

    private final List<Specification<Event>> specifications;

    public GebeurtenisApiDelegateService(List<Specification<Event>> specifications) {
        this.specifications = specifications;
    }

    @Override
    public ResponseEntity<PublishEvent200Response> publishEvent(Event event) {
        if (event == null) return new ResponseEntity<>(HttpStatus.BAD_REQUEST);

        boolean anySatisfied = specifications.stream().anyMatch(spec -> spec.isSatisfiedBy(event));

        String result = anySatisfied ? UUID.randomUUID().toString() : "Invalid event type or specification mismatch";

        PublishEvent200Response response = new PublishEvent200Response();
        response.setId(result);
        return new ResponseEntity<>(response, HttpStatus.OK);
    }
}

File: src/main/kotlin/nl/brp/api/gebeurtenissen/web/api/GebeurtenisApiDelegateService.kt
package nl.brp.api.gebeurtenissen.web.api

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import java.util.*

@Service
class GebeurtenisApiDelegateService(
        private val specifications: List<Specification<Event>>,
        private val commandHandler: GebeurtenisCommandHandler
    ) : GebeurtenisApiDelegate {
    override fun publishEvent(event: Event?): ResponseEntity<PublishEvent200Response> {
        if (event == null) return ResponseEntity(HttpStatus.BAD_REQUEST)

        val anySatisfied = specifications.stream().anyMatch { spec: Specification<Event> -> spec.isSatisfiedBy(event) }

        if(!anySatisfied) return ResponseEntity(HttpStatus.BAD_REQUEST)

        val result = UUID.randomUUID().toString()

        val command = PublishEventCommand(
            id = result,
            event = event
        )

        commandHandler.handle(command)

        val response = PublishEvent200Response(id = result)
        return ResponseEntity(response, HttpStatus.OK)
    }
}